### PR TITLE
[clang] remove ClassScopeFunctionSpecializationDecl

### PR DESCRIPTION
**Pull request by sdkrystian**
This removes the `ClassScopeFunctionSpecializationDecl` `Decl` node, and instead uses `DependentFunctionTemplateSpecializationInfo` to handle such declarations. `DependentFunctionTemplateSpecializationInfo` is also changed to store a `const ASTTemplateArgumentListInfo*` to be more in line with `FunctionTemplateSpecializationInfo`. 

This also changes `FunctionDecl::isFunctionTemplateSpecialization` to return `true` for dependent specializations, and `FunctionDecl::getTemplateSpecializationKind`/`FunctionDecl::getTemplateSpecializationKindForInstantiation` to return `TSK_ExplicitSpecialization` for non-friend dependent specializations (the same behavior as dependent class scope `ClassTemplateSepcializationDecl` & `VarTemplateSepcializationDecl`).

### CODE REVIEW

**Comment by cor3ntin** [comment]
@erichkeane Opinions?

**Review by erichkeane — COMMENTED** [review]
I guess my first question is 'why'?  My second is 'if this is unnecessary, why was it here in the first place?`.  I'd like some level of historical analysis here/perhaps the knowledge of @zygoloid.

**Comment by sdkrystian** [comment]
@erichkeane 
> I guess my first question is 'why'? 

Per the comment you mentioned, it is redundant. Switching to `DependentFunctionTemplateSpecializationInfo` also results in constructs such as:

```
template<typename T>
struct A {
  template<>
  void f();
};
```

being diagnosed prior to instantiation. Class scope explicit specializations currently exist in their own broken little corner of the frontend, so unifying their representation in the AST would be the first step towards properly fixing them. 

> My second is 'if this is unnecessary, why was it here in the first place?`. 

`ClassScopeFunctionSpecializationDecl` was originally added in [this commit](https://github.com/llvm/llvm-project/commit/00c7e6ceb1826baab50c02f8547e97bfcaf9641c) to "parse MSVC standard C++ headers, MFC and ATL code." Back then, class scope explicit specializations were a Microsoft extension, but since [CWG727](https://wg21.link/cwg727) they are valid ISO C++. As for why this was not originally implemented using `DependentFunctionTemplateSpecializationInfo`, I cannot say.

**Review by erichkeane — COMMENTED** [review]
I didn't spend much time in Serialization, so hopefully someone else can take a look, but in general I think I'm warming toward this.

File: clang/lib/Sema/SemaDecl.cpp
Comment (sdkrystian) [reply clang/lib/Sema/SemaDecl.cpp:10491]: I opted to not fix it in this commit to minimize unrelated changes. We _should_ be doing this check if `!isFriend` (since we already diagnose storage class specifiers on friend declarations elsewhere). Furthermore, the check for inconsistent storage class should happen in `CheckFunctionTemplateSpecialization` (so we check during the instantiation of dependent specializations, once the primary template is known).

File: clang/include/clang/AST/RecursiveASTVisitor.h
Comment (sdkrystian) [reply clang/include/clang/AST/RecursiveASTVisitor.h:2147]: I originally added this comment because it changes behavior (i.e. explicit template arguments for dependent friends were not previously traversed, but are traversed with this change). Upon a second examination, I don't think the comment is necessary, and the correct behavior is to traverse the explicit arguments (as we would traverse them for a non-dependent specialization).

File: clang-tools-extra/clang-tidy/modernize/UseEmplaceCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/modernize/UseEmplaceCheck.cpp:70]:
```
  if (const FunctionTemplateDecl* PrimaryTemplate = Node.getMethodDecl()->getPrimaryTemplate())
    return Node.getNumArgs() == PrimaryTemplate->getTemplatedDecl()->getNumParams();
```

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp
Comment (shafik) [inline clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:4076]:
```suggestion
        MD, /*TemplateParams=*/nullptr,
```

Comment (shafik) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:4076]: To be consistent with [bugprone-argument-comment](https://clang.llvm.org/extra/clang-tidy/checks/bugprone/argument-comment.html)

File: clang/lib/Serialization/ASTReaderDecl.cpp
Comment (shafik) [inline clang/lib/Serialization/ASTReaderDecl.cpp:1007]: It looks like this should be `uint64_t`, although it does not seem to do what the "label says" it does.

File: clang/lib/Serialization/ASTWriterDecl.cpp
Comment (shafik) [inline clang/lib/Serialization/ASTWriterDecl.cpp:646]: `getNumTemplates()` returns `unsigned` while `size()` AFAICT returns `size_t` which is not the same type. Is this ok?

**Review by shafik — COMMENTED** [review]
Mostly minor fixes and or questions.

File: clang/lib/Serialization/ASTWriterDecl.cpp
Comment (sdkrystian) [reply clang/lib/Serialization/ASTWriterDecl.cpp:646]: `DependentFunctionTemplateSpecializationInfo` stores the number of candidates in an `unsigned` member, so it should be fine.

File: clang/lib/Serialization/ASTReaderDecl.cpp
Comment (sdkrystian) [reply clang/lib/Serialization/ASTReaderDecl.cpp:1007]: @shafik  The "convention" here seems to be to use the type of the member being deserialized (see lines 1082 & 1122). The `UnresolvedSet` is only used to pass the candidates to `DependentFunctionTemplateSpecializationInfo::Create`; the actual candidates are stored as trailing objects & the number of candidates is stored in `DependentFunctionTemplateSpecializationInfo::NumCandidates` (which has the type `unsigned`).

**Comment by sdkrystian** [comment]
Ping @erichkeane & @shafik -- I left comments responding to your feedback

**Comment by erichkeane** [comment]
I don't have any comments, though someone with knowledge of teh clang-tools-extra stuff would be nice to take a look, as well as someone familiar with the serialization.  @AaronBallman .

ALSO, this might need a release note and does this need to be mailed to the 'breaking changes' vendor list since it is significantly going to change the AST?

**Review by AaronBallman — COMMENTED** [review]
The changes should definitely have a release note so that downstreams can be aware of the changes. This isn't really a potentially breaking change, but we should alert clang-vendors in case a downstream actually cares.

**Comment by sdkrystian** [comment]
@AaronBallman 
> Can you be sure to file an issue or add a test case + FIXME comment so we don't lose track of the needed follow-up work, or are you planning to do that work immediately after this lands?

I will address that once this gets merged

**Comment by AaronBallman** [comment]
CC @llvm/clang-vendors for awareness of the AST changing to remove a node.

File: clang-tools-extra/clangd/SemanticHighlighting.cpp
Comment (AaronBallman) [reply clang-tools-extra/clangd/SemanticHighlighting.cpp:756]: Ahh, thank you, that makes sense!

**Comment by sdkrystian** [comment]
Updated the release notes. I didn't go into exhaustive detail, but I can add more if anyone here deems it necessary @erichkeane @AaronBallman

File: clang/docs/ReleaseNotes.rst
Comment (AaronBallman) [inline clang/docs/ReleaseNotes.rst:51]:
```suggestion
to store candidate primary templates and explicit template arguments. This should not impact users
of Clang as a compiler, but it may break assumptions in Clang-based tools iterating over the AST.
```

**Review by AaronBallman — APPROVED** [review]
LGTM aside from a minor suggestion with the release note.

**Comment by AaronBallman** [comment]
Do you need someone to land these changes on your behalf?

**Comment by sdkrystian** [comment]
@AaronBallman I don't have write access, so that would be appreciated 🙂. I was putting off rebasing incase there were more responses to my comments, but I presume that there won't be so I will get it rebased now

**Comment by erichkeane** [comment]
If Aaron doesn't get to it, feel free to ping me on Monday.  Its too late on a Friday for me to be able to merge this for you, but monday morning would be fine for me.

**Pull request merged**
